azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -345,6 +482,148 @@ def test_connect_failure(self, mocker, stage, op, arbitrary_exception):
+    ):
+        stage.pipeline_root.connected = False
+
+        stage.run_op(op)
+
+        # complete the connection
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+        stage.pipeline_root.connected = True
+        connect_op.complete()
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 2
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that we got a new connect operation
+        assert stage.send_op_down.call_count == 3
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+
+    @pytest.mark.it(
+        "Runs the operation a second time if the stage needs to reconnect after the op fails"
+    )
+    def test_failure_and_disconnect_resubmit_operation(
+        self, mocker, stage, op, arbitrary_exception
+    ):
+        stage.pipeline_root.connected = False
+
+        stage.run_op(op)
+
+        # complete the connection
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+        stage.pipeline_root.connected = True
+        connect_op.complete()
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 2
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that we got a new connect operation
+        assert stage.send_op_down.call_count == 3
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+
+        # complete the connect op
+        stage.pipeline_root.connected = True
+        connect_op.complete()
+
+        # make sure our op got sent down again
+        assert stage.send_op_down.call_count == 4
+        assert stage.send_op_down.call_args[0][0] is op
+
+    @pytest.mark.it("Fails the op if the op fails and the transport is still connected")
+    def test_failure_still_connected(self, mocker, stage, op, arbitrary_exception):
+        callback = op.callback_stack[0]
+        stage.pipeline_root.connected = False
+
+        stage.run_op(op)
+
+        # complete the connection
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+        stage.pipeline_root.connected = True
+        connect_op.complete()
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 2
+        assert stage.send_op_down.call_args[0][0] is op
```

Comment:
I'm not a huge fan of the fact that this test doens't do any kind of check for the first time the call goes down. Essentially, we're asserting it's been called twice, but only in order to validate the second one.

I understand that the first call happens as part of the setup, but it still might be good to check that it's already been called once as part of the precondition before the action that sends it into a second call. That's generally the pattern I try to use.

Not a huge deal though tbh.